text/0076-entity-slice-validation.md

```diff
@@ -0,0 +1,440 @@
+
+## Detailed design
+
+### Entity dereferences
+
+An "Entity Dereference" is any operation that requires directly accessing an entity’s data. The following is an exhaustive list of entity dereferencing operations
+
+1. `e1 in e2` dereferences `e1` (but not `e2`)
+2. `e1.foo` dereferences `e1` iff `e1` is an entity
+3. `e1 has foo` dereferences `e1` iff `e1` is an entity
+
+Note that testing equality *is not* a dereferencing operation. Therefore expressions like `e1 == e2`, `e1.contains(e2)` and `e3 in e2` do not dereference `e1` or `e2`.
+
+Also note that by “dereference” above, we are only referring to the given operation, not about any operations that might be executed to evaluate subexpressions. For example, while `e1 in e2` only dereferences `e1` (per rule 1 above), if `e2` is `principal.foo`, then of course evaluating `e2` dereferences `principal` (per rule 2 above).
+
+### Entity roots
+
+The variables `principal`, `action`, and `resource` are *entity roots*. (Entity literals are *not* entity roots, and operations on such literals are limited, as described below.) They directly name entities from a request that could be dereferenced in a policy. A `context` is not an entity root directly; rather, any attributes it (transitively) contains that have entity type are considered roots. For example, consider the following schema:
+
+```
+entity User {
+    manager: User,
+    age: Long,
+};
+action getDetails appliesTo {
+    principal: User,
+    resource: User,
+    context: {
+        admin: User,
+        building: { location: Long, ITDeptHead: User },
+        foo: { inner: User },
+        bar: { inner: User }
+    }
+};
+```
+
+Here, `principal`, `action`, and `resource` are entity roots (as usual), as are:
+
+* `context.admin` 
+* `context.building.ITDeptHead` 
+* `context.foo.inner` , and 
+* `context.bar.inner`
+
+Note that `context`,  `context.building`, and  `context.building.location` are *not* entity roots because they do not have entity type. Moreover, `context.admin.manager` is not an entity root because `manager` is not an attribute of `context` itself.
+
+### Dereference levels
+
+During validation, we extend entity types with a *level*, which is either `∞` or a natural number. We call these *leveled entity types*. The level expresses how many entity dereferences are permitted starting from an expression having that type. Levels on types are only used internally; policy/schema writers never see them. When the level is `∞`, validation is unchanged from the status quo.
+
+### Specifying the level
+
+The maximum level that policies may validate at is specified in the schema. The benefit of this is that policy writers can see the expected form of policies in one place. For example, we could extend our schema above as follows:
+
+```
+level = 1;
```

Comment:
I like this syntax (and prefer the info here as opposed to an argument to validation). Minor implementation question: if this syntax is used in one schema fragment, does it automatically apply to all other schema fragments? Or do the other schema fragments also have to explicitly use the same level?

Reply:
This is a great point.

I would propose that each fragment must declare its own level, and it's an error if the levels do not agree.  Alternately, if we take Alternative C (per-entity levels), it would be natural to allow different levels in each fragment, where the level in each fragment applies to all entity types declared in that fragment.

Reply:
Proposal: When combining fragments, at most one can have a `level` specification, else it's an error. 

This works when you are importing a bunch of common and entity type defs into a single app, where the app specifies the level.

Reply:
This makes more sense than my initial proposal when you consider the possibility that you have library-like fragments of entity definitions (or common-type definitions, etc). There's no reason library authors should (have to or want to) specify a level, it seems like purely a consideration for the app.